### PR TITLE
proposal: NRI Mode Resource Management

### LINKED ISSUE #1277: [proposal] NRI on Koordinator
### Discussed in https://github.com/koordinator-sh/koordinator/discussions/1276

<div type='discussions-op-text'>

<sup>Originally posted by **saintube** May  8, 2023</sup>
In this talk, we discuss why and how the [Node Resource Interface (NRI)](https://github.com/containerd/nri) could be used in Koordinator.

### Overview

> ref: https://mp.weixin.qq.com/s/gwww7Is2_lG7m20wdAeWSg

There are several approaches to extending the Kubernetes CRI (Container Runtime Interface) to manage container resources (e.g. adjust the cpuset of a container).

- Default
  - kubelet -- CRI Request -> CRI Runtime (containerd/cri-o) -- OCI Spec -> OCI-compatible runtime (runc/kata) -> containers
- CRI Proxy (e.g. koordlet + runtime-proxy)
  - kubelet -- CRI Request **-> CRI Proxy -- CRI Request (hooked) ->** CRI Runtime -- OCI Spec -> OCI compatible runtime -> containers
- Standalone (e.g. koordlet)
  - kubelet -- CRI Request -> CRI Runtime -- OCI Spec -> OCI compatible runtime -> containers
  - **kubelet ---> Node Agent -> CRI Runtime / containers**

approaches

koordlet-runtimehooks

Instead of modifying the CRI request, NRI offers us a new way to manager container resources by adjusting OCI-spec.

nri-approach

### NRI API

> The Runtime service is the public interface runtimes expose for NRI plugins. All requests on this interface are initiated by the plugin. The interface provides functions for
> - initiating plugin registration
> - requesting unsolicited updates to containers
> 
> The Plugin service is the public interface NRI uses to interact with plugins. All requests on this interface are initiated by NRI/the runtime. The interface provides functions for
> - configuring the plugin
> - getting initial list of already existing pods and containers
> - hooking the plugin into pod/container lifecycle events
> - shutting down the plugin

#### Objects

Pod

```go
// https://github.com/containerd/nri/blob/v0.3.0/pkg/api/api.pb.go#L1015
type PodSandbox struct {
	// ...
	Id             string            `protobuf:"bytes,1,opt,name=id,proto3" json:"id,omitempty"`
	Name           string            `protobuf:"bytes,2,opt,name=name,proto3" json:"name,omitempty"`
	Uid            string            `protobuf:"bytes,3,opt,name=uid,proto3" json:"uid,omitempty"`
	Namespace      string            `protobuf:"bytes,4,opt,name=namespace,proto3" json:"namespace,omitempty"`
	Labels         map[string]string `protobuf:"bytes,5,rep,name=labels,proto3" json:"labels,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	Annotations    map[string]string `protobuf:"bytes,6,rep,name=annotations,proto3" json:"annotations,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	RuntimeHandler string            `protobuf:"bytes,7,opt,name=runtime_handler,json=runtimeHandler,proto3" json:"runtime_handler,omitempty"`
	Linux          *LinuxPodSandbox  `protobuf:"bytes,8,opt,name=linux,proto3" json:"linux,omitempty"`
	Pid            uint32            `protobuf:"varint,9,opt,name=pid,proto3" json:"pid,omitempty"` // for NRI v1 emulation
}

type LinuxPodSandbox struct {
	// ...
	PodOverhead  *LinuxResources   `protobuf:"bytes,1,opt,name=pod_overhead,json=podOverhead,proto3" json:"pod_overhead,omitempty"`
	PodResources *LinuxResources   `protobuf:"bytes,2,opt,name=pod_resources,json=podResources,proto3" json:"pod_resources,omitempty"`
	CgroupParent string            `protobuf:"bytes,3,opt,name=cgroup_parent,json=cgroupParent,proto3" json:"cgroup_parent,omitempty"`
	CgroupsPath  string            `protobuf:"bytes,4,opt,name=cgroups_path,json=cgroupsPath,proto3" json:"cgroups_path,omitempty"` // for NRI v1 emulation
	Namespaces   []*LinuxNamespace `protobuf:"bytes,5,rep,name=namespaces,proto3" json:"namespaces,omitempty"`                      // for NRI v1 emulation
	Resources    *LinuxResources   `protobuf:"bytes,6,opt,name=resources,proto3" json:"resources,omitempty"`                        // for NRI v1 emulation
}
```

Container

```go
type Container struct {
	// ...
	Id           string            `protobuf:"bytes,1,opt,name=id,proto3" json:"id,omitempty"`
	PodSandboxId string            `protobuf:"bytes,2,opt,name=pod_sandbox_id,json=podSandboxId,proto3" json:"pod_sandbox_id,omitempty"`
	Name         string            `protobuf:"bytes,3,opt,name=name,proto3" json:"name,omitempty"`
	State        ContainerState    `protobuf:"varint,4,opt,name=state,proto3,enum=nri.pkg.api.v1alpha1.ContainerState" json:"state,omitempty"`
	Labels       map[string]string `protobuf:"bytes,5,rep,name=labels,proto3" json:"labels,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	Annotations  map[string]string `protobuf:"bytes,6,rep,name=annotations,proto3" json:"annotations,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	Args         []string          `protobuf:"bytes,7,rep,name=args,proto3" json:"args,omitempty"`
	Env          []string          `protobuf:"bytes,8,rep,name=env,proto3" json:"env,omitempty"`
	Mounts       []*Mount          `protobuf:"bytes,9,rep,name=mounts,proto3" json:"mounts,omitempty"`
	Hooks        *Hooks            `protobuf:"bytes,10,opt,name=hooks,proto3" json:"hooks,omitempty"`
	Linux        *LinuxContainer   `protobuf:"bytes,11,opt,name=linux,proto3" json:"linux,omitempty"`
	Pid          uint32            `protobuf:"varint,12,opt,name=pid,proto3" json:"pid,omitempty"` // for NRI v1 emulation
}

type LinuxContainer struct {
	// ...
	Namespaces  []*LinuxNamespace `protobuf:"bytes,1,rep,name=namespaces,proto3" json:"namespaces,omitempty"`
	Devices     []*LinuxDevice    `protobuf:"bytes,2,rep,name=devices,proto3" json:"devices,omitempty"`
	Resources   *LinuxResources   `protobuf:"bytes,3,opt,name=resources,proto3" json:"resources,omitempty"`
	OomScoreAdj *OptionalInt      `protobuf:"bytes,4,opt,name=oom_score_adj,json=oomScoreAdj,proto3" json:"oom_score_adj,omitempty"`
	CgroupsPath string            `protobuf:"bytes,5,opt,name=cgroups_path,json=cgroupsPath,proto3" json:"cgroups_path,omitempty"`
}
```

#### Managed Resources

LinuxResources

```go
type LinuxResources struct {
	// ...
	Memory         *LinuxMemory         `protobuf:"bytes,1,opt,name=memory,proto3" json:"memory,omitempty"`
	Cpu            *LinuxCPU            `protobuf:"bytes,2,opt,name=cpu,proto3" json:"cpu,omitempty"`
	HugepageLimits []*HugepageLimit     `protobuf:"bytes,3,rep,name=hugepage_limits,json=hugepageLimits,proto3" json:"hugepage_limits,omitempty"`
	BlockioClass   *OptionalString      `protobuf:"bytes,4,opt,name=blockio_class,json=blockioClass,proto3" json:"blockio_class,omitempty"`
	RdtClass       *OptionalString      `protobuf:"bytes,5,opt,name=rdt_class,json=rdtClass,proto3" json:"rdt_class,omitempty"`
	Unified        map[string]string    `protobuf:"bytes,6,rep,name=unified,proto3" json:"unified,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	Devices        []*LinuxDeviceCgroup `protobuf:"bytes,7,rep,name=devices,proto3" json:"devices,omitempty"` // for NRI v1 emulation
}
```

LinuxDevices

```go
type LinuxDevice struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Path     string            `protobuf:"bytes,1,opt,name=path,proto3" json:"path,omitempty"`
	Type     string            `protobuf:"bytes,2,opt,name=type,proto3" json:"type,omitempty"`
	Major    int64             `protobuf:"varint,3,opt,name=major,proto3" json:"major,omitempty"`
	Minor    int64             `protobuf:"varint,4,opt,name=minor,proto3" json:"minor,omitempty"`
	FileMode *OptionalFileMode `protobuf:"bytes,5,opt,name=file_mode,json=fileMode,proto3" json:"file_mode,omitempty"`
	Uid      *OptionalUInt32   `protobuf:"bytes,6,opt,name=uid,proto3" json:"uid,omitempty"`
	Gid      *OptionalUInt32   `protobuf:"bytes,7,opt,name=gid,proto3" json:"gid,omitempty"`
}
```

### How-To Access

There are two ways to access NRI:

1. Pre-Registered Plugins
    - Configs prepared in the directory `/opt/nri/plugins`.
2. External Plugins
    - Plugins communicate and register with NRI Adaption via `/var/run/nri/nri.sock`.
    - Runtime can configure to disable External Plugins.

### Access NRI

Add a Koord NRI Plugin for

1. Get OCI specs from the NRI Adaption.
2. Transform the OCI spec into internal protocols (i.e. PodContext, ContainerContext) to re-use existing runtime hook plugins.

### Workflow

... --> CRI Runtime - OCI Spec -> NRI Adaption --> Koord NRI Plugin - OCI Spec (adjusted) -> ...

nri-on-koordinator

### vs. Koord Runtime Proxy

runtime-proxy

- Advantages
  - Adopt the upstream standard of the container runtime.
  - Pluggable deployed than the Runtime Proxy.
- Challenges
  - Need to upgrade container >= 1.7.0.
  - Fault tolerance: koordlet expects some plugin failures to be tolerated.
    - Koordlet: (PluginFailPolicy) https://github.com/koordinator-sh/koordinator/blob/main/pkg/koordlet/runtimehooks/hooks/hooks.go#L79
    - NRI: (`if FatalError(err) { plugin.close() } else abort`) https://github.com/containerd/nri/blob/v0.3.0/pkg/adaptation/adaptation.go#L290

### Use Cases

- Fine-grained CPU scheduling, CPU Suppress (cpuset mode) -> adjust containers' cpuset
- CPU Suppress (cfsquota mode) -> adjust containers' cfs quota
- Batch Resource -> manage cgroup resources of BE pods
- Memory QoS -> set pod/container's memory.min, memory.low, memory.high
- LLC / Memory Bandwidth Isolation -> control containers' RDT group
- ...

</div>

### PR DESCRIPTION
**Pull request by kangclzjc**
### Ⅰ. Describe what this PR does

Koordinator as a QoS based scheduling system for hybrid workloads orchestration on Kubernetes and its runtime hooks support two working [modes](https://github.com/koordinator-sh/koordinator/blob/main/docs/design-archive/koordlet-runtime-hooks.md) for different scenarios: `Standalone` and `Proxy`. However, both of them have some [constraints](https://shimo.im/docs/m4kMLdgO1LIma9qD) in 2023/02/07 meeting discussion. We'd like to integrate NRI framework to address these constraints based on community recommend mechanism.

Different from standalone and proxy mode, Koodlet will start an NRI plugin to subscribe pod/container lifecycle events from container runtime (e.g. containerd, crio), and then koordlet NRI plugin will call runtime hooks to adjust pod resources or OCI spec. 

### Ⅱ. Does this pull request fix one issue?

fixes #1277 

### Ⅲ. Describe how to verify it

Deploy pods which need hooks adjustment resources to check the behaviour of pod.
### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

**Comment by hormes** [comment]
As we discussed, embracing the community mechanism has always been a Koordinator principle. This is a good exploration, and I hope we can work together to promote the maturity of NRI.

**Comment by saintube** [comment]
@kangclzjc Thank you for your contribution. The proposal is concise and well-organized. Let’s continue discussing some details at the weekly meeting.
BTW, please check the [DCO]((https://github.com/koordinator-sh/koordinator/pull/1366/checks?check_run_id=14097269044)), which requires that commits are signed-off.

**Comment by kangclzjc** [comment]
> @kangclzjc Thank you for your contribution. The proposal is concise and well-organized. Let’s continue discussing some details at the weekly meeting. BTW, please check the DCO, which requires that commits are signed-off.

Thanks @saintube. Add signed-off now.

File: docs/proposals/20230608-nri-mode-resource-management.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func (p *nriServer) RunPodSandbox(pod *api.PodSandbox) error {
+	podCtx.FromNri(pod)
+	RunHooks(...)
+	podCtx.NriDone()
+}
+
+func (p *nriServer) CreateContainer(pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+    containerCtx.FromNri(pod, container)
```
Comment (saintube) [inline docs/proposals/20230608-nri-mode-resource-management.md:128]: please check the indentations

File: docs/proposals/20230608-nri-mode-resource-management.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+### Risks and Mitigations
+
+## Upgrade Strategy
+
+- Need to upgrade containerd to 1.7.0+ or CRIO to 1.25.0+
+- Need to enable NRI
+
+### Test Plan [optional]
+
```
Comment (saintube) [inline docs/proposals/20230608-nri-mode-resource-management.md:169]: As we discussed in the biweekly meeting, please add an *Alternative* section for comparisons with the runtime-proxy mode.

File: docs/proposals/20230608-nri-mode-resource-management.md (hunk 1 above)
Comment (kangclzjc) [reply docs/proposals/20230608-nri-mode-resource-management.md:129]: Fixed.

File: docs/proposals/20230608-nri-mode-resource-management.md (hunk 2 above)
Comment (kangclzjc) [reply docs/proposals/20230608-nri-mode-resource-management.md:185]: Yes, added. Compare runtime-proxy with NRI briefly now.

Comment (kangclzjc) [reply docs/proposals/20230608-nri-mode-resource-management.md:195]: > As we discussed in the biweekly meeting, please add an _Alternative_ section for comparisons with the runtime-proxy mode.

Also added the comparisons with standalone mode.

**Comment by hormes** [comment]
Can you describe the difference in execution timing between NRI and RuntimeProxy hook points, which is very important for subsequent iterations?

**Comment by kangclzjc** [comment]
> Can you describe the difference in execution timing between NRI and RuntimeProxy hook points, which is very important for subsequent iterations?

Yes, there is a little difference in execution timing between them. The biggest difference is RuntimeProxy call koordlet hooks between kubelet and containerd. However, NRI will call koodlet hooks in containerd, that means containerd still could do something before or after containerd call NRI plugin(koordlet hooks).

**Review by hormes — APPROVED** [review]

**Pull request merged**